react/task2/src/components/CounterParent.js

```diff
@@ -0,0 +1,169 @@
+import React from "react";
+import Button from '@material-ui/core/Button'
+import Grid from '@material-ui/core/Grid'
+import CounterContainer from "./CounterContainer";
+import { CounterContainerStyle, TextStyle, GridStyle } from '../views/Counter/style'
+import scrollToComponent from 'react-scroll-to-component';
+
+const initialState = [{ count: 0, id: 1 }]
+
+class CounterParent extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            counters: initialState
+        };
+
+        this.handleIncrement = this.handleIncrement.bind(this);
+        this.handleDecrement = this.handleDecrement.bind(this);
+        this.handleReset = this.handleReset.bind(this);
+
+        console.log("constructor:CounterParent");
+    }
+
+    componentDidMount() {
+        console.log("componentDidMount:CounterParent");
+    }
+
+    shouldComponentUpdate(nextProps, nextState) {
+        console.log("shouldComponentUpdate:CounterParent");
+        return (nextState.counters !== this.state.counters);
+    }
+
+    getSnapshotBeforeUpdate(prevProps, prevState) {
+        console.log("getSnapshotBeforeUpdate:CounterParent")
+        if (prevState.counters.length > 30 && prevState.counters.length < this.state.counters.length) {
+            const element = document.querySelector("#counters:last-child");
+            return element;
+        }
+        return null;
+    }
+
+    componentDidUpdate(prevProps, prevState, snapshot) {
+        console.log("componentDidUpdate:CounterParent")
+        if (snapshot !== null) {
+            console.log(snapshot);
+            scrollToComponent(snapshot, {
+                offset: 1000,
+                align: 'bottom',
+                duration: 1500
+            });
+        }
+    }
+
+    componentWillUnmount() {
+        console.log("componentWillUnmount:CounterParent");
+    }
+
+    randomInteger = (min, max) => {
+        var rand = min - 0.5 + Math.random() * (max - min + 1)
+        rand = Math.round(rand);
+        return rand;
+    }
+
+    handleIncrement = (id) => {
+        let newCounters = this.state.counters.map((counter) => {
```

Comment:
When you are changing state, you should change the reference to object/array in the state. In your code, you changed array, but the reference to it is still the same. It's wrong. You could use the spread operator`(...)` to create a copy of the old array and push a new array element into it.